blueprint/core/blueprint_EcmascriptEngine.cpp

```diff
@@ -252,8 +252,101 @@ namespace blueprint
             return result;
         }
 
+        struct TimeoutFunctionManager : private juce::MultiTimer
+        {
+            ~TimeoutFunctionManager() override {
+                for(const auto &[id, timer] : timeoutFunctions)
+                    stopTimer(id);
+            }
+
+            juce::var clearTimeout(const int id)
+            {
+                stopTimer(id);
+                const auto f = timeoutFunctions.find(id);
+                if(f != timeoutFunctions.cend())
+                    timeoutFunctions.erase(f);
+                return juce::var();
+            }
+
+            int newTimeout(const juce::var::NativeFunction f, const int timeoutMillis, const std::vector<juce::var>&& args, const bool repeats=false)
+            {
+                static int nextId = 0;
+                timeoutFunctions.emplace(nextId, TimeoutFunction(f, std::move(args), repeats));
+                startTimer(nextId, timeoutMillis);
+                return nextId++;
+            }
+
+            void timerCallback(int id) override
+            {
+                const auto f = timeoutFunctions.find(id);
+                if(f != timeoutFunctions.cend())
+                {
+                    const auto cb = f->second;
+                    std::invoke(cb.f, juce::var::NativeFunctionArgs(juce::var(), cb.args.data(), static_cast<int>(cb.args.size())));
+                    if(!cb.repeats)
+                    {
+                        stopTimer(id);
+                        timeoutFunctions.erase(f);
+                    }
+                }
+            }
+
+            private:
+                struct TimeoutFunction
+                {
+                    TimeoutFunction(const juce::var::NativeFunction _f, const std::vector<juce::var> &&_args, const bool _repeats=false)
+                    : f(_f), args(std::move(_args)), repeats(_repeats) {}
+
+                    const juce::var::NativeFunction f;
+                    std::vector<juce::var> args;
+                    const bool repeats;
+                };
+
+                std::map<int, TimeoutFunction> timeoutFunctions;
+        };
+
+        // IsSetter is true for setTimeout / setInterval
+        // and false for clearTimeout / clearInterval
+        template <bool IsSetter = false, bool Repeats = false, typename MethodType>
```

Comment:
I toyed with an enum template parameter to make it clearer in calls what this bool refers to, but it just seemed a little excessive.  There are only ever going to be 4 calls to this thing anyway so a comment will do.

Reply:
Also, I could only port so much of the cool stuff from https://github.com/nick-thompson/blueprint/commit/45584e748ce9db27c9e4ddf328f6343298d24935 over, purely because of the nature of having one type of call that wants to copy an unknown number of additional arguments over etc.  We've probably ended up at about the same code-weight here.  Nonetheless, I do slightly prefer this result.